File: ios/Oriveo/Oriveo/Core/Providers/GrokSubscriptionRuntime.swift
import Foundation

enum GrokSubscriptionRuntime {
    struct Prepared: Sendable, Equatable {
        let accessToken: String
        let context: GrokSubscriptionRequestContext
        let didRefresh: Bool
    }

    static func prepare(
        providerID: UUID,
        uid: String = AppSessionStore.activeUID,
        availability: GrokSubscriptionAvailability? = nil,
        client: GrokSubscriptionOAuthClient = GrokSubscriptionOAuthClient(),
        now: Date = Date()
    ) async -> Result<Prepared, GrokSubscriptionError> {
        let resolvedAvailability = availability ?? MetadataClient.shared.syncGrokSubscriptionAvailability()
        guard case let .available(config) = resolvedAvailability else {
            return .failure(.configurationUnavailable)
        }

        guard let stored = GrokSubscriptionCredentialStore.load(providerID: providerID, uid: uid) else {
            return .failure(.unauthorized)
        }

        guard let chatURL = config.chatURL else {
            return .failure(.configurationUnavailable)
        }
        let context = GrokSubscriptionRequestContext(
            chatURL: chatURL,
            responsesURL: config.responsesURL,
            requiredHeaders: config.requiredHeaders,
            transport: (GrokSubscriptionAuthConfig.transportKind(apiBackend: config.apiBackend) ?? .openaiChat).rawValue
        )

        guard stored.needsRefresh(now: now) else {
            return .success(Prepared(accessToken: stored.accessToken, context: context, didRefresh: false))
        }

        guard let refreshToken = stored.refreshToken else {
            return .failure(.unauthorized)
        }

        do {
            let refreshed = try await client.refreshTokens(config: config, refreshToken: refreshToken)
            GrokSubscriptionCredentialStore.save(refreshed, providerID: providerID, uid: uid)
            return .success(Prepared(accessToken: refreshed.accessToken, context: context, didRefresh: true))
        } catch {
            if let expiresAt = stored.expiresAt, now < expiresAt {
                return .success(Prepared(accessToken: stored.accessToken, context: context, didRefresh: false))
            }
            return .failure(.unauthorized)
        }
    }

    static func persist(tokens: GrokSubscriptionTokens, providerID: UUID, uid: String = AppSessionStore.activeUID) {
        GrokSubscriptionCredentialStore.save(tokens, providerID: providerID, uid: uid)
    }

    static func disconnect(
        providerID: UUID,
        uid: String = AppSessionStore.activeUID,
        client: GrokSubscriptionOAuthClient = GrokSubscriptionOAuthClient()
    ) async {
        if case let .available(config) = MetadataClient.shared.syncGrokSubscriptionAvailability(),
           let stored = GrokSubscriptionCredentialStore.load(providerID: providerID, uid: uid) {
            await client.revoke(config: config, token: stored.accessToken)
        }
        GrokSubscriptionCredentialStore.delete(providerID: providerID, uid: uid)
    }
}

extension GrokSubscriptionError {
    var userFacingMessage: String { L10n.tr(userFacingMessageKey, table: .providers) }

    var userFacingMessageKey: String {
        switch self {
        case .clientVersionRejected, .configurationUnavailable:
            return "Grok subscription sign-in is temporarily unavailable while we update it. You can connect with an API key instead."
        case .subscriptionNotEligible:
            return "Your xAI account's current plan doesn't allow using the Grok subscription in third-party apps."
        case .unauthorized:
            return "Your Grok sign-in has expired. Please authorize again."
        case .quotaExhausted:
            return "You've used up this period's Grok subscription quota. It will resume after the next reset."
        case .codeExpired:
            return "The authorization code expired. Please start again."
        case .accessDenied:
            return "Authorization was declined."
        case .authorizationPending, .slowDown:
            return "Waiting for authorization in your browser…"
        case .transport, .upstream:
            return "Couldn't reach Grok. Check your connection and try again."
        }
    }


    var requiresConfigRefresh: Bool {
        self == .clientVersionRejected
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/provider/grok/GrokSubscriptionRuntime.kt
package ai.oriveo.community.core.provider.grok

import ai.oriveo.community.core.security.SecureKeyStore
import kotlinx.serialization.json.Json

/**
 * Facade over OAuth credential storage for Grok subscription sign-in.
 *
 * It only converts between serialized form and encrypted storage; when and why a write
 * happens is decided by [GrokSubscriptionRuntime]. The underlying namespace is separate from
 * the apiKey slot, see [SecureKeyStore.saveSubscriptionCredential].
 */
class GrokSubscriptionCredentialStore(
    private val secureKeyStore: SecureKeyStore,
) {
    private val json = Json { ignoreUnknownKeys = true }

    fun save(accountId: String, providerID: String, tokens: GrokSubscriptionTokens) {
        secureKeyStore.saveSubscriptionCredential(
            accountId = accountId,
            providerID = providerID,
            payload = json.encodeToString(GrokSubscriptionTokens.serializer(), tokens),
        )
    }

    fun load(accountId: String, providerID: String): GrokSubscriptionTokens? {
        val raw = secureKeyStore.loadSubscriptionCredential(accountId, providerID)
            ?.takeIf { it.isNotBlank() } ?: return null
        return runCatching {
            json.decodeFromString(GrokSubscriptionTokens.serializer(), raw)
        }.getOrNull()
    }

    fun delete(accountId: String, providerID: String) {
        secureKeyStore.deleteSubscriptionCredential(accountId, providerID)
    }
}

/**
 * Credential preparation before an outbound subscription request: fetch the current access
 * token, renew it when needed, and assemble the outbound context.
 *
 * Kept separate from [GrokSubscriptionOAuthClient] (pure networking) and
 * [GrokSubscriptionCredentialStore] (pure storage): this is the layer that strings the two
 * together around "what has to happen before one send", and the single place that decides
 * between renewing, demanding a fresh sign-in, and simply letting the request through.
 */
class GrokSubscriptionRuntime(
    private val credentialStore: GrokSubscriptionCredentialStore,
    private val oauthClient: GrokSubscriptionOAuthClient,
    private val availabilityProvider: () -> GrokSubscriptionAvailability,
    private val now: () -> Long = System::currentTimeMillis,
) {
    data class Prepared(
        val accessToken: String,
        val context: GrokSubscriptionRequestContext,
        /**
         * Whether a renewal happened on this call. If it did, the refresh token in storage
         * has been rotated, so a caller still holding an older snapshot has to re-read it.
         */
        val didRefresh: Boolean,
    )

    sealed class PrepareResult {
        data class Success(val prepared: Prepared) : PrepareResult()
        data class Failure(val error: GrokSubscriptionError) : PrepareResult()
    }

    /**
     * Assembles everything one subscription request needs.
     *
     * Failures always return a specific meaning rather than a generic error, because the
     * caller decides which sentence the user sees from it: "please authorize again" and
     * "your subscription tier does not cover this" call for completely different next steps.
     */
    suspend fun prepare(accountId: String, providerID: String): PrepareResult {
        val availability = availabilityProvider()
        val config = (availability as? GrokSubscriptionAvailability.Available)?.config
            // Switched off by the kill switch, or absent from the config: stop sending.
            // Degradation notices for already-connected instances are the UI's job; all
            // this does is guarantee we never keep hitting upstream with a config that
            // may already be invalid.
            ?: return PrepareResult.Failure(GrokSubscriptionError.ConfigurationUnavailable)

        val stored = credentialStore.load(accountId, providerID)
            ?: return PrepareResult.Failure(GrokSubscriptionError.Unauthorized)

        val context = GrokSubscriptionRequestContext(
            chatUrl = config.chatUrl,
            responsesUrl = config.responsesUrl,
            requiredHeaders = config.requiredHeaders,
            transport = GrokSubscriptionAuthResolver.transportKind(config.apiBackend)
                ?: ai.oriveo.community.core.provider.transport.TransportKind.OpenAIResponses.wireValue,
        )

        val timestamp = now()
        if (!stored.needsRefresh(timestamp)) {
            return PrepareResult.Success(Prepared(stored.accessToken, context, didRefresh = false))
        }

        // Expired with no refresh token in hand: the only way out is a fresh authorization.
        val refreshToken = stored.refreshToken
            ?: return PrepareResult.Failure(GrokSubscriptionError.Unauthorized)

        return try {
            val refreshed = oauthClient.refreshTokens(config, refreshToken)
            // xAI rotates refresh_token on every refresh, so the whole record is written
            // back; otherwise the next renewal would present an already-revoked token.
            credentialStore.save(accountId, providerID, refreshed)
            PrepareResult.Success(Prepared(refreshed.accessToken, context, didRefresh = true))
        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e
        } catch (_: Exception) {
            // The refresh failed but the old token has not truly expired yet, so let it
            // through: needsRefresh fires 5 minutes early and the old token is still valid
            // inside that window. Kicking the user back to sign-in over one network blip
            // would be far too blunt.
            val expiresAt = stored.expiresAt
            if (expiresAt != null && timestamp < expiresAt) {
                PrepareResult.Success(Prepared(stored.accessToken, context, didRefresh = false))
            } else {
                PrepareResult.Failure(GrokSubscriptionError.Unauthorized)
            }
        }
    }

    /** Persists tokens after a first or repeat authorization succeeds. */
    fun persist(accountId: String, providerID: String, tokens: GrokSubscriptionTokens) {
        credentialStore.save(accountId, providerID, tokens)
    }

    fun load(accountId: String, providerID: String): GrokSubscriptionTokens? =
        credentialStore.load(accountId, providerID)

    /**
     * Disconnects the subscription: best-effort revoke upstream first, then delete locally.
     *
     * A failed revoke does not block the local delete - the user pressed Disconnect, so this
     * device should not keep a usable token lying around either way.
     */
    suspend fun disconnect(accountId: String, providerID: String) {
        val config = (availabilityProvider() as? GrokSubscriptionAvailability.Available)?.config
        val stored = credentialStore.load(accountId, providerID)
        if (config != null && stored != null) {
            oauthClient.revoke(config, stored.accessToken)
        }
        credentialStore.delete(accountId, providerID)
    }
}

/**
 * Translates a subscription-route failure into a user-visible Provider error.
 *
 * Intermediate states (pending / slow_down) and transport or unclassified upstream errors
 * are deliberately not in this table: the former are absorbed by the state machine itself,
 * and the latter stay on the generic network error path rather than being crammed into one
 * of the four subscription categories to fake a precise diagnosis.
 */
fun GrokSubscriptionError.toProviderServiceError(): ai.oriveo.community.core.model.ProviderServiceError =
    when (this) {
        is GrokSubscriptionError.ClientVersionRejected,
        is GrokSubscriptionError.ConfigurationUnavailable,
        -> ai.oriveo.community.core.model.ProviderServiceError.GrokSubscription(
            ai.oriveo.community.core.model.GrokSubscriptionFailureReason.ClientVersionRejected,
            detail = toString(),
        )

        is GrokSubscriptionError.SubscriptionNotEligible ->
            ai.oriveo.community.core.model.ProviderServiceError.GrokSubscription(
                ai.oriveo.community.core.model.GrokSubscriptionFailureReason.NotEligible,
                detail = toString(),
            )

        is GrokSubscriptionError.Unauthorized ->
            ai.oriveo.community.core.model.ProviderServiceError.GrokSubscription(
                ai.oriveo.community.core.model.GrokSubscriptionFailureReason.Expired,
                detail = toString(),
            )

        is GrokSubscriptionError.QuotaExhausted ->
            ai.oriveo.community.core.model.ProviderServiceError.GrokSubscription(
                ai.oriveo.community.core.model.GrokSubscriptionFailureReason.QuotaExhausted,
                detail = toString(),
            )

        is GrokSubscriptionError.Transport ->
            ai.oriveo.community.core.model.ProviderServiceError.Network(detail)

        is GrokSubscriptionError.Upstream ->
            ai.oriveo.community.core.model.ProviderServiceError.Upstream(status, body)

        is GrokSubscriptionError.AuthorizationPending,
        is GrokSubscriptionError.SlowDown,
        is GrokSubscriptionError.CodeExpired,
        is GrokSubscriptionError.AccessDenied,
        -> ai.oriveo.community.core.model.ProviderServiceError.Upstream(0, toString())
    }
